### DISCUSSION TITLE
Prevent fzf from changing my selection without a keystroke?

### DISCUSSION THREAD

**Discussion by falsifian** [junegunn/fzf#3088] [Q&A] [answered]
Here's a common issue I face:
* I run (basically) `find | fzf` and type `foo`.
* The directory `bar/foo`, which is the one I wanted, gets selected. `find` is still running, though...
* I decide to press enter. But just before I press enter, `find` comes across `baz/foo`, and `fzf` decides that's a better match.
* Result: I select `baz/foo` when I meant to select `bar/foo`.

My ideal solution to this would be: fzf never changes the selection unless I type something. In the above example, fzf would stay on `bar/foo`, but then e.g. if I typed space it would update it to the better match. This prevents the selection being changed between when I decide to press enter and actually press enter.

Is there an option for that? Or can someone suggest a different solution? If I submitted a pull request to add such an option, is that likely to be accepted?

**Comment by junegunn** [comment, marked as answer]
There is `--sync` option which blocks fzf until the whole input is ready.

> My ideal solution to this would be: fzf never changes the selection unless I type something.

I don't think this is going to turn out great. For example, you start fzf, and you quickly type in the query string, but the possible candidates from the `find` command are not ready. So you get empty or suboptimal results. You have to delete and re-type the query string just to trigger the search.

**Reply by falsifian** [reply]
> There is `--sync` option which blocks fzf until the whole input is ready.

Thanks, good to know, though not exactly what I'm looking for.

> > My ideal solution to this would be: fzf never changes the selection unless I type something.
> 
> I don't think this is going to turn out great. For example, you start fzf, and you quickly type in the query string, but the possible candidates from the `find` command are not ready. So you get empty or suboptimal results. You have to delete and re-type the query string just to trigger the search.

I disagree here. Two things to keep in mind:
* There's nothing to stop fzf from *showing* new results as they come in. I just don't want a new result to be selected without my deliberate action.
* There's no need to delete and retype. Just press space once: that counts as a keystroke, and shouldn't change the results.

To be clear, this is what I'm imagining:
* I run `find | fzf --no-bait-and-switch`.
* I'm looking for `baz/foo`, so I type `foo`. It hasn't been output by find yet, so fzf selects `bar/foo` instead.
* Half a second later, I notice 'bar/foo` has appeared in the results. Maybe it's highlighted in a special colour to indicate that it's the real first result, and fzf is just aching to select it for me the moment I press a key.
* I press space, see that `baz/foo` is selected, then press enter.

**Reply by junegunn** [reply]
Thanks for the suggestion, but I'm probably not going to add the option as it feels a bit too complicated to me. It may seem simple and straightforward in this particular, minimal example, but there are more things to consider UI-wise as the number of matched items before the "space" trigger and the number of new items after it (some ranked higher, some ranked lower than the current match) increase. You also have to decide how to handle the current cursor position (which may not be on the first page of the list) without a jarring effect. Having said that, I will test the feature if you send a pull request, I may change my mind in the end. But even if I don't accept it, you can still use your version, so why not?

**Reply by falsifian** [reply]
Understood, thanks for your thoughts. I might give it a shot at some point.
